In [117]:
df = pd.read_csv('nba_logreg2.csv', sep = ";")

In [118]:
df_continuas = df.drop(columns=['Name', 'TARGET_5Yrs'])
df_continuas.head()

GP   MIN  PTS  FGM  FGA   FG%  3P Made  3PA   3P%  FTM  FTA   FT%  OREB  \
0  36 27.40 7.40 2.60 7.60 34.70     0.50 2.10 25.00 1.60 2.30 69.90  0.70   
1  35 26.90 7.20 2.00 6.70 29.60     0.70 2.80 23.50 2.60 3.40 76.50  0.50   
2  74 15.30 5.20 2.00 4.70 42.20     0.40 1.70 24.40 0.90 1.30 67.00  0.50   
3  58 11.60 5.70 2.30 5.50 42.60     0.10 0.50 22.60 0.90 1.30 68.90  1.00   
4  48 11.50 4.50 1.60 3.00 52.40     0.00 0.10  0.00 1.30 1.90 67.40  1.00   

   DREB  REB  AST  STL  BLK  TOV  
0  3.40 4.10 1.90 0.40 0.40 1.30  
1  2.00 2.40 3.70 1.10 0.50 1.60  
2  1.70 2.20 1.00 0.50 0.30 1.00  
3  0.90 1.90 0.80 0.60 0.10 1.00  
4  1.50 2.50 0.30 0.30 0.40 0.80

In [119]:
# Definir el modelo y el grid de hiperparámetros
model = RandomForestClassifier(random_state=42)
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}


In [120]:
# Configurar y ajustar GridSearchCV
grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=5, n_jobs=-1, verbose=2, scoring='accuracy')
grid_search.fit(X_train, y_train)

# Obtener y mostrar los mejores hiperparámetros
best_params = grid_search.best_params_
print("Mejores hiperparámetros:", best_params)

# Entrenar el modelo con los mejores hiperparámetros
best_model = RandomForestClassifier(**best_params, random_state=42)
best_model.fit(X_train, y_train)

# Evaluar el modelo
y_pred = best_model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Precisión del modelo:", accuracy)
print(classification_report(y_test, y_pred))


Fitting 5 folds for each of 108 candidates, totalling 540 fits
Mejores hiperparámetros: {'max_depth': 10, 'min_samples_leaf': 4, 'min_samples_split': 2, 'n_estimators': 200}
Precisión del modelo: 0.6766917293233082
              precision    recall  f1-score   support

         0.0       0.57      0.44      0.50        97
         1.0       0.72      0.81      0.76       169

    accuracy                           0.68       266
   macro avg       0.65      0.63      0.63       266
weighted avg       0.66      0.68      0.67       266



Interpretacion de resultados: 
El modelo de Random Forest, con los hiperparámetros optimizados, muestra un rendimiento razonable con una precisión general del 67.67%. La clase 1 se predice con mayor precisión y recall en comparación con la clase 0. El desbalance de clases podría estar afectando el rendimiento, y se podrían explorar técnicas adicionales para manejar este desbalance y mejorar el modelo.


Resumen de lo que hicimos anteriormente: 

Se definen las columnas numéricas y la columna objetivo.
Se verifica que todas las columnas necesarias existen en el DataFrame.
Se seleccionan las columnas numéricas y la columna objetivo.
Estandarizar los Datos:

Se estandarizan las columnas numéricas utilizando StandardScaler para que tengan media 0 y desviación estándar 1.
Dividir los Datos:

Ahora se dividen los datos en conjuntos de entrenamiento y prueba usando train_test_split.
Optimización de Hiperparámetros:

Entonces definimos un modelo RandomForestClassifier y una cuadrícula de hiperparámetros (param_grid).
Luego configura y ajusta GridSearchCV para encontrar la mejor combinación de hiperparámetros.
Entrenar el Modelo con los Mejores Hiperparámetros:

Se entrena el modelo utilizando los mejores hiperparámetros obtenidos de GridSearchCV.
Evaluar el Modelo:

En resumen se realizan predicciones en el conjunto de prueba. Ademas se calcula y muestra la precisión del modelo. Para finalmente generar un informe de clasificación (classification_report) que incluye métricas como precisión, recall y F1-score.